File: frontend/src/lib/api/websocket.ts
import { io, Socket } from 'socket.io-client'
import type { WSJobUpdate } from '@/types'

const WS_URL = process.env.NEXT_PUBLIC_WS_URL || 'ws://localhost:8000'

type EventHandler<T = any> = (data: T) => void

class WebSocketClient {
  private socket: Socket | null = null
  private listeners: Map<string, Set<EventHandler>> = new Map()
  private reconnectAttempts = 0
  private maxReconnectAttempts = 5
  private reconnectDelay = 1000

  connect(): void {
    if (this.socket?.connected) return

    this.socket = io(WS_URL, {
      transports: ['websocket', 'polling'],
      reconnection: true,
      reconnectionDelay: this.reconnectDelay,
      reconnectionAttempts: this.maxReconnectAttempts,
    })

    this.socket.on('connect', () => {
      console.log('WebSocket connected')
      this.reconnectAttempts = 0
      this.emit('connection', { connected: true })
    })

    this.socket.on('disconnect', () => {
      console.log('WebSocket disconnected')
      this.emit('connection', { connected: false })
    })

    this.socket.on('connect_error', (error) => {
      console.error('WebSocket connection error:', error)
      this.reconnectAttempts++

      if (this.reconnectAttempts >= this.maxReconnectAttempts) {
        console.error('Max reconnection attempts reached')
        this.emit('error', { message: 'Failed to connect to server' })
      }
    })

    // Job-specific events
    this.socket.on('job:update', (data: WSJobUpdate) => {
      this.emit('job:update', data)
      this.emit(`job:${data.job_id}`, data)
    })

    this.socket.on('job:completed', (data: WSJobUpdate) => {
      this.emit('job:completed', data)
      this.emit(`job:${data.job_id}:completed`, data)
    })

    this.socket.on('job:error', (data: WSJobUpdate) => {
      this.emit('job:error', data)
      this.emit(`job:${data.job_id}:error`, data)
    })

    // Batch events
    this.socket.on('batch:update', (data) => {
      this.emit('batch:update', data)
    })
  }

  disconnect(): void {
    if (this.socket) {
      this.socket.disconnect()
      this.socket = null
    }
    this.listeners.clear()
  }

  // Subscribe to specific job updates
  subscribeToJob(jobId: string): void {
    if (!this.socket) this.connect()
    this.socket?.emit('subscribe:job', { job_id: jobId })
  }

  // Unsubscribe from job updates
  unsubscribeFromJob(jobId: string): void {
    this.socket?.emit('unsubscribe:job', { job_id: jobId })
  }

  // Subscribe to batch updates
  subscribeToBatch(batchId: string): void {
    if (!this.socket) this.connect()
    this.socket?.emit('subscribe:batch', { batch_id: batchId })
  }

  // Generic event listener
  on<T = any>(event: string, handler: EventHandler<T>): void {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, new Set())
    }
    this.listeners.get(event)!.add(handler)
  }

  // Remove event listener
  off<T = any>(event: string, handler: EventHandler<T>): void {
    const handlers = this.listeners.get(event)
    if (handlers) {
      handlers.delete(handler)
      if (handlers.size === 0) {
        this.listeners.delete(event)
      }
    }
  }

  // Emit to local listeners
  private emit(event: string, data: any): void {
    const handlers = this.listeners.get(event)
    if (handlers) {
      handlers.forEach((handler) => handler(data))
    }
  }

  // Check connection status
  isConnected(): boolean {
    return this.socket?.connected || false
  }
}

// Singleton instance
export const wsClient = new WebSocketClient()

// React hook-friendly API
export function useWebSocket() {
  return {
    connect: () => wsClient.connect(),
    disconnect: () => wsClient.disconnect(),
    subscribeToJob: (jobId: string) => wsClient.subscribeToJob(jobId),
    unsubscribeFromJob: (jobId: string) => wsClient.unsubscribeFromJob(jobId),
    subscribeToBatch: (batchId: string) => wsClient.subscribeToBatch(batchId),
    on: <T = any>(event: string, handler: EventHandler<T>) => wsClient.on(event, handler),
    off: <T = any>(event: string, handler: EventHandler<T>) => wsClient.off(event, handler),
    isConnected: () => wsClient.isConnected(),
  }
}

File: api/app/websocket.py
"""
WebSocket Manager for Real-time Job Updates

Provides real-time bidirectional communication between backend and frontend
using Socket.IO for job status updates, progress tracking, and batch updates.
"""

import socketio
from typing import Dict, Set, Optional, Any
from datetime import datetime
import logging

# Configure logging
logger = logging.getLogger(__name__)

# Create Socket.IO server instance
sio = socketio.AsyncServer(
    async_mode='asgi',
    cors_allowed_origins='*',  # Configure per environment
    logger=True,
    engineio_logger=True
)

# Track active subscriptions
# job_id -> set of session IDs
job_subscriptions: Dict[str, Set[str]] = {}
# batch_id -> set of session IDs
batch_subscriptions: Dict[str, Set[str]] = {}
# session_id -> set of subscribed job_ids
session_jobs: Dict[str, Set[str]] = {}
# session_id -> set of subscribed batch_ids
session_batches: Dict[str, Set[str]] = {}


class WebSocketManager:
    """
    Manages WebSocket connections and event broadcasting.
    """

    @staticmethod
    async def broadcast_job_update(
        job_id: str,
        status: str,
        progress: float,
        message: str
    ) -> None:
        """
        Broadcast job update to all subscribed clients.

        Args:
            job_id: Job identifier
            status: Current job status
            progress: Progress percentage (0-100)
            message: Status message
        """
        data = {
            'job_id': job_id,
            'status': status,
            'progress': progress,
            'message': message,
            'timestamp': datetime.utcnow().isoformat()
        }

        # Emit to all subscribers of this specific job
        if job_id in job_subscriptions:
            for session_id in job_subscriptions[job_id]:
                await sio.emit('job:update', data, room=session_id)
                logger.debug(f"Sent job:update to {session_id} for job {job_id}")

    @staticmethod
    async def broadcast_job_completed(
        job_id: str,
        message: str = "Job completed successfully"
    ) -> None:
        """
        Broadcast job completion to all subscribed clients.

        Args:
            job_id: Job identifier
            message: Completion message
        """
        data = {
            'job_id': job_id,
            'status': 'DELIVERED',
            'progress': 100,
            'message': message,
            'timestamp': datetime.utcnow().isoformat()
        }

        if job_id in job_subscriptions:
            for session_id in job_subscriptions[job_id]:
                await sio.emit('job:completed', data, room=session_id)
                await sio.emit('job:update', data, room=session_id)
                logger.info(f"Sent job:completed to {session_id} for job {job_id}")

    @staticmethod
    async def broadcast_job_error(
        job_id: str,
        error_message: str
    ) -> None:
        """
        Broadcast job error to all subscribed clients.

        Args:
            job_id: Job identifier
            error_message: Error description
        """
        data = {
            'job_id': job_id,
            'status': 'ABORTED',
            'progress': 0,
            'message': f"Error: {error_message}",
            'timestamp': datetime.utcnow().isoformat()
        }

        if job_id in job_subscriptions:
            for session_id in job_subscriptions[job_id]:
                await sio.emit('job:error', data, room=session_id)
                await sio.emit('job:update', data, room=session_id)
                logger.error(f"Sent job:error to {session_id} for job {job_id}")

    @staticmethod
    async def broadcast_batch_update(
        batch_id: str,
        total: int,
        completed: int,
        status: str,
        message: str
    ) -> None:
        """
        Broadcast batch processing update to all subscribed clients.

        Args:
            batch_id: Batch identifier
            total: Total jobs in batch
            completed: Number of completed jobs
            status: Batch status
            message: Status message
        """
        data = {
            'batch_id': batch_id,
            'total': total,
            'completed': completed,
            'progress': (completed / total * 100) if total > 0 else 0,
            'status': status,
            'message': message,
            'timestamp': datetime.utcnow().isoformat()
        }

        if batch_id in batch_subscriptions:
            for session_id in batch_subscriptions[batch_id]:
                await sio.emit('batch:update', data, room=session_id)
                logger.debug(f"Sent batch:update to {session_id} for batch {batch_id}")


# Socket.IO Event Handlers

@sio.event
async def connect(sid: str, environ: dict):
    """
    Handle client connection.

    Args:
        sid: Session ID
        environ: Connection environment
    """
    logger.info(f"Client connected: {sid}")
    session_jobs[sid] = set()
    session_batches[sid] = set()

    # Send connection confirmation
    await sio.emit('connection', {'connected': True, 'session_id': sid}, room=sid)


@sio.event
async def disconnect(sid: str):
    """
    Handle client disconnection. Clean up subscriptions.

    Args:
        sid: Session ID
    """
    logger.info(f"Client disconnected: {sid}")

    # Clean up job subscriptions
    if sid in session_jobs:
        for job_id in session_jobs[sid]:
            if job_id in job_subscriptions:
                job_subscriptions[job_id].discard(sid)
                if not job_subscriptions[job_id]:
                    del job_subscriptions[job_id]
        del session_jobs[sid]

    # Clean up batch subscriptions
    if sid in session_batches:
        for batch_id in session_batches[sid]:
            if batch_id in batch_subscriptions:
                batch_subscriptions[batch_id].discard(sid)
                if not batch_subscriptions[batch_id]:
                    del batch_subscriptions[batch_id]
        del session_batches[sid]


@sio.on('subscribe:job')
async def handle_subscribe_job(sid: str, data: dict):
    """
    Subscribe client to job updates.

    Args:
        sid: Session ID
        data: {'job_id': str}
    """
    job_id = data.get('job_id')
    if not job_id:
        logger.warning(f"subscribe:job called without job_id by {sid}")
        return

    # Add to subscriptions
    if job_id not in job_subscriptions:
        job_subscriptions[job_id] = set()
    job_subscriptions[job_id].add(sid)

    if sid not in session_jobs:
        session_jobs[sid] = set()
    session_jobs[sid].add(job_id)

    logger.info(f"Client {sid} subscribed to job {job_id}")

    # Acknowledge subscription
    await sio.emit('subscribed', {'job_id': job_id}, room=sid)


@sio.on('unsubscribe:job')
async def handle_unsubscribe_job(sid: str, data: dict):
    """
    Unsubscribe client from job updates.

    Args:
        sid: Session ID
        data: {'job_id': str}
    """
    job_id = data.get('job_id')
    if not job_id:
        return

    # Remove from subscriptions
    if job_id in job_subscriptions:
        job_subscriptions[job_id].discard(sid)
        if not job_subscriptions[job_id]:
            del job_subscriptions[job_id]

    if sid in session_jobs:
        session_jobs[sid].discard(job_id)

    logger.info(f"Client {sid} unsubscribed from job {job_id}")


@sio.on('subscribe:batch')
async def handle_subscribe_batch(sid: str, data: dict):
    """
    Subscribe client to batch updates.

    Args:
        sid: Session ID
        data: {'batch_id': str}
    """
    batch_id = data.get('batch_id')
    if not batch_id:
        logger.warning(f"subscribe:batch called without batch_id by {sid}")
        return

    # Add to subscriptions
    if batch_id not in batch_subscriptions:
        batch_subscriptions[batch_id] = set()
    batch_subscriptions[batch_id].add(sid)

    if sid not in session_batches:
        session_batches[sid] = set()
    session_batches[sid].add(batch_id)

    logger.info(f"Client {sid} subscribed to batch {batch_id}")

    # Acknowledge subscription
    await sio.emit('subscribed', {'batch_id': batch_id}, room=sid)


@sio.on('unsubscribe:batch')
async def handle_unsubscribe_batch(sid: str, data: dict):
    """
    Unsubscribe client from batch updates.

    Args:
        sid: Session ID
        data: {'batch_id': str}
    """
    batch_id = data.get('batch_id')
    if not batch_id:
        return

    # Remove from subscriptions
    if batch_id in batch_subscriptions:
        batch_subscriptions[batch_id].discard(sid)
        if not batch_subscriptions[batch_id]:
            del batch_subscriptions[batch_id]

    if sid in session_batches:
        session_batches[sid].discard(batch_id)

    logger.info(f"Client {sid} unsubscribed from batch {batch_id}")


# Export singleton instance
ws_manager = WebSocketManager()
